pkg/scheduler/plugins/cpunormalization/resource_allocation.go

```diff
@@ -0,0 +1,147 @@
+limitations under the License.
+*/
+
+package cpunormalization
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/klog/v2"
+	"k8s.io/kubernetes/pkg/scheduler/apis/config"
+	"k8s.io/kubernetes/pkg/scheduler/framework"
+	schedutil "k8s.io/kubernetes/pkg/scheduler/util"
+
+	"github.com/koordinator-sh/koordinator/apis/extension"
+)
+
+// resourceToWeightMap contains resource name and weight.
+type resourceToWeightMap map[corev1.ResourceName]int64
+
+// scorer is decorator for resourceAllocationScorer
+type scorer func(args *config.NodeResourcesFitArgs) *resourceAllocationScorer
+
+// resourceAllocationScorer contains information to calculate resource allocation score.
+type resourceAllocationScorer struct {
+	Name string
+	// used to decide whether to use Requested or NonZeroRequested for
+	// cpu and memory.
+	useRequested        bool
+	scorer              func(requested, allocatable resourceToValueMap) int64
+	resourceToWeightMap resourceToWeightMap
+}
+
+// resourceToValueMap is keyed with resource name and valued with quantity.
+type resourceToValueMap map[corev1.ResourceName]int64
+
+// score will use `scorer` function to calculate the score.
+func (r *resourceAllocationScorer) score(
+	pod *corev1.Pod,
+	nodeInfo *framework.NodeInfo) (int64, *framework.Status) {
+	node := nodeInfo.Node()
+	if node == nil {
+		return 0, framework.NewStatus(framework.Error, "node not found")
+	}
+	if r.resourceToWeightMap == nil {
+		return 0, framework.NewStatus(framework.Error, "resources not found")
+	}
+
+	nRatio, err := extension.GetCPUNormalizationRatio(node)
+	if err != nil {
+		return 0, framework.AsStatus(err)
+	}
+	if nRatio == -1 {
+		// CPU normalization disabled is semantically equal to CPU normalization ratio 1
+		nRatio = 1
+	}
+
+	requested := make(resourceToValueMap)
+	allocatable := make(resourceToValueMap)
+	for resource := range r.resourceToWeightMap {
+		alloc, req := r.calculateResourceAllocatableRequest(nodeInfo, pod, resource, nRatio)
+		if alloc != 0 {
+			// Only fill the extended resource entry when it's non-zero.
+			allocatable[resource], requested[resource] = alloc, req
+		}
+	}
+
+	score := r.scorer(requested, allocatable)
+
+	klog.V(10).InfoS("Listing internal info for allocatable resources, requested resources and score with CPU normalization",
+		"pod", klog.KObj(pod), "node", klog.KObj(node), "resourceAllocationScorer", r.Name,
+		"allocatableResource", allocatable, "requestedResource", requested, "resourceScore", score,
+	)
+
+	return score, nil
+}
+
+// calculateResourceAllocatableRequest returns 2 parameters:
+// - 1st param: quantity of allocatable resource on the node.
+// - 2nd param: aggregated quantity of requested resource on the node.
+// Note: if it's an extended resource, and the pod doesn't request it, (0, 0) is returned.
+func (r *resourceAllocationScorer) calculateResourceAllocatableRequest(nodeInfo *framework.NodeInfo, pod *corev1.Pod, resource corev1.ResourceName, nRatio float64) (int64, int64) {
```

Comment:
Plugins just handle CPU resources, so we can simplify functionality and remove unused code.

Reply:
This only applies to the filter logic I think. We should keep the complete score logic of `NodeResourcesFit` to make sure our score differs from `NodeResourcesFit` only under CPU normalization cases. If we only socre CPU, our score would be very different from which `NodeResourcesFit` gives.

Reply:
Ok. got it. You are right.